mlflow/utils/rest_utils.py

```diff
@@ -77,6 +92,36 @@ def http_request(
     Returns:
         requests.Response object.
     """
+    if host_creds.databricks_workspace_client:
+        try:
+            if method == "GET":
+                extra_kwargs = {"query": kwargs["params"]}
+            else:
+                extra_kwargs = {"body": kwargs["json"]}
+
+            raw_response = host_creds.databricks_workspace_client.api_client.do(
+                method=method,
+                path=endpoint,
+                headers=extra_headers,
+                raw=True,
+                **extra_kwargs,
+            )
+            return raw_response["contents"]._response
```

Comment:
Accessing `_response` seems scary. why do we need it?

Reply:
This is for getting original `requests.Response` object. The reason is, in MLFlow deployment, because we support LLM streaming, it calls `requests.Response.iterlines` method, so return the original `requests.Response` object avoids modifying MLFlow deployment streaming prediction code.

I checked databricks-sdk code, `_response` attribute exists in all versions >= 0.20, we can ask `databricks-sdk` team to make it as a public attribute.

Reply:
> databricks-sdk code, _response attribute exists in all versions >= 0.20

This doesn't mean we can use it. The `_` prefix means "don't touch it".

Reply:
wait do we need to use `databricks_workspace_client.api_client` for oauth to work?

Reply:
`databricks_workspace_client.api_client` is not the only way, however, it should be the simplest way.

I saw sql team set up oauth by writing some code by themselves: https://github.com/databricks/databricks-sql-python/blob/main/src/databricks/sql/auth/authenticators.py  but this way we have to maintain more code and it is error-prone.

On the other hand, databricks-sdk supports many other authentication ways besides OAuth, if we can make mlflow authentication using databricks-sdk, the other authentication ways we don't need extra effort to support.

Reply:
> This doesn't mean we can use it. The _ prefix means It's private. Don't touch it.

`databricks_workspace_client.api_client.do` is a low-level api that just wraps original `requests.Response` object. We can negotiate with them to add public interface to get original `Response` object, I think technically it is reasonable.

Reply:
Can we control how we retry the request if we use `databricks_workspace_client.api_client`?

Reply:
> Can we control how we retry the request (or other options for HTTP requests) if we use databricks_workspace_client.api_client?

No. databricks-sdk design goal is to automatically pick optimal value for these configs, so they don't expose param for them.

Reply:
> This doesn't mean we can use it. The _ prefix means It's private. Don't touch it.

Avoid using `_response` is doable too, I can add some wrapping code over `databricks_workspace_client.api_client` public interfaces to implement equivalent functionality of `response.iter_lines` streaming.

Reply:
> No. databricks-sdk design goal is to automatically pick optimal value for these configs

can you elaborate on "automatically"?

Reply:
I need to check databricks-sdk source code for details and then get back to you.

Reply:
`databricks-sdk` exposes a config `retry_timeout_seconds`, official doc:
> (Integer) Number of seconds to keep retrying HTTP requests. Default is 300 (5 minutes). (from https://github.com/databricks/databricks-sdk-py)

For other HTTP controlling params, **they are not exposed**, internally, the logic for retry backoff is somewhat coarse. Retryable errors are defined in ApiClient._is_retriable , and the retry respects the Retry-After header if present in a 429 or 503 response. The backoff is linear, starting at 1 second and increasing by 1 second up to 10 seconds. Jitter of between 0-1 second is added for each retry.

Reply:
Proposal:

Add a environmental var config `DATABRICKS_SDK_HTTP_RETRY_TIMEOUT_SECONDS`, for requests invoked by databricks-sdk

Reply:
Config `DATABRICKS_ENDPOINT_HTTP_RETRY_TIMEOUT_SECONDS` was added.